File: docs/snippets/all/archetypes/transform3d_column_updates.rs
//! Update a transform over time, in a single operation.
//!
//! This is semantically equivalent to the `transform3d_row_updates` example, albeit much faster.

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let rec =
        rerun::RecordingStreamBuilder::new("rerun_example_transform3d_column_updates").spawn()?;

    rec.set_time_sequence("tick", 0);
    rec.log(
        "box",
        &[
            &rerun::Boxes3D::from_half_sizes([(4.0, 2.0, 1.0)])
                .with_fill_mode(rerun::FillMode::Solid) as &dyn rerun::AsComponents,
            &rerun::TransformAxes3D::new(10.0),
        ],
    )?;

    let translations = (0..100).map(|t| [0.0, 0.0, t as f32 / 10.0]);
    let rotations = (0..100)
        .map(|t| truncated_radians((t * 4) as f32))
        .map(|rad| rerun::RotationAxisAngle::new([0.0, 1.0, 0.0], rerun::Angle::from_radians(rad)));

    let ticks = rerun::TimeColumn::new_sequence("tick", 1..101);
    rec.send_columns(
        "box",
        [ticks],
        rerun::Transform3D::default()
            .with_many_translation(translations)
            .with_many_rotation_axis_angle(rotations)
            .columns_of_unit_batches()?,
    )?;

    Ok(())
}

fn truncated_radians(deg: f32) -> f32 {
    ((deg.to_radians() * 1000.0) as i32) as f32 / 1000.0
}

File: docs/snippets/all/archetypes/transform3d_column_updates.cpp
//! Update a transform over time, in a single operation.
//!
//! This is semantically equivalent to the `transform3d_row_updates` example, albeit much faster.

#include <cmath>
#include <numeric>
#include <vector>

#include <rerun.hpp>

float truncated_radians(int deg) {
    auto degf = static_cast<float>(deg);
    const auto pi = 3.14159265358979323846f;
    return static_cast<float>(static_cast<int>(degf * pi / 180.0f * 1000.0f)) / 1000.0f;
}

int main(int argc, char* argv[]) {
    const auto rec = rerun::RecordingStream("rerun_example_transform3d_column_updates");
    rec.spawn().exit_on_failure();

    rec.set_time_sequence("tick", 0);
    rec.log(
        "box",
        rerun::Boxes3D::from_half_sizes({{4.f, 2.f, 1.0f}}).with_fill_mode(rerun::FillMode::Solid),
        rerun::TransformAxes3D(10.0)
    );

    std::vector<std::array<float, 3>> translations;
    std::vector<rerun::RotationAxisAngle> rotations;
    for (int t = 0; t < 100; t++) {
        translations.push_back({0.0f, 0.0f, static_cast<float>(t) / 10.0f});
        rotations.push_back(rerun::RotationAxisAngle(
            {0.0f, 1.0f, 0.0f},
            rerun::Angle::radians(truncated_radians(t * 4))
        ));
    }

    std::vector<int64_t> ticks(100);
    std::iota(ticks.begin(), ticks.end(), 1);

    rec.send_columns(
        "box",
        rerun::TimeColumn::from_sequence("tick", ticks),
        rerun::Transform3D()
            .with_many_translation(translations)
            .with_many_rotation_axis_angle(rotations)
            .columns()
    );
}
